### PR TITLE
Fix training with ctc loss

### PR DESCRIPTION
**Pull request by ljj7975**
CTC loss was in the code but seems like it was broken for a long time.
I have a feeling that it has never been supported.

- blank label was incorrectly setup (first n are for positive labels, n+1 for out of vocab word, n+2 for the blank)
- sequence inference requires smoothing to be off as detection happens per word

requested by #42

### CODE REVIEW

**Review by edwinzhng — APPROVED** [review]

File: howl/data/tokenize.py [hunk 1]
```diff
@@ -101,14 +102,24 @@ def decode(self, ids: List[int]) -> str:
 
     def encode(self, transcript: str) -> List[int]:
         encoded_output = []
+        # append a space at the beginning to catch vocab starting with a space
+        # append a space to add the last negative label
+        transcript = " " + transcript + " "
```
Comment (daemon) [inline howl/data/tokenize.py:107]: Do `f' {transcript} '` instead.

Comment (daemon) [reply howl/data/tokenize.py:107]: And does the codebase use single quotes or double quotes? I would keep the style consistent.

File: howl/model/inference.py [hunk 2]
```diff
@@ -228,10 +226,11 @@ def __init__(self, sample_rate: int, *args, blank_idx: int = 0, **kwargs):
     def infer(self, audio_data: torch.Tensor) -> bool:
         delta_ms = int(audio_data.size(-1) / self.sample_rate * 1000)
         self.std = self.std.to(audio_data.device)
-        scores = self.model(self.zmuv(self.std(audio_data.unsqueeze(0))), None)
-        scores = F.softmax(scores, -1).squeeze(1)
+        scores = self.model(self.zmuv(self.std(audio_data.unsqueeze(0))), None) # [num_frames x 1 (batch size) x num_labels]
```
Comment (daemon) [inline howl/model/inference.py:229]: Two spaces between the ) and #.

File: howl/run/train.py [hunk 3]
```diff
@@ -197,20 +199,26 @@ def do_evaluate():
...
+                scores = F.log_softmax(scores, -1) # [num_frames x batch_size x num_labels]
                 lengths = torch.tensor([model.compute_length(x.item()) for x in lengths]).to(device)
                 loss = criterion(scores, batch.labels, lengths, batch.label_lengths)
             optimizer.zero_grad()
             model.zero_grad()
             loss.backward()
             optimizer.step()
             pbar.set_postfix(dict(loss=f'{loss.item():.3}'))
-            writer.add_scalar('Training/Loss', loss.item(), epoch_idx)
+            total_loss += loss
```
Comment (daemon) [inline howl/run/train.py:210]: You want to do
```
with torch.no_grad():
  total_loss += loss
```
to avoid accumulating the loss gradient.

**Review by daemon — CHANGES_REQUESTED** [review]

**Comment by daemon** [comment]
Can you fix the conflicts then we can merge?

**Comment by ljj7975** [comment]
@daemon I have found that demo.py does not work with ctc version. 
I will fix it in a separate MR
